### DISCUSSION TITLE
Wrap up the official release v3.1.2 (or v3.2.0)

### DISCUSSION THREAD

**Discussion by mokun** [mars-sim/mars-sim#366] [General]
@bevans2000,

What do you consider are the **must-fixes** prior to the official release `v3.1.2` ?

Last time we just already moved the milestones once for the next two releases a couple of month further out. 

As a result, the `v3.1.2` is due by **Jul 1, 2021**. 

So it's two weeks from now. See [milestones](https://github.com/mars-sim/mars-sim/milestones).

Q: what date can we agree upon within next week as the last day for freezing the changes so that we can push out the v3.1.2 ? 

This way, we can project wrapping up our work earlier. If it doesn't work at the last minute, we can postpone that a few more days and still have more leeways.

Right afterwards, I'd say we can rightfully jump to v3.2.0 since we've made tons of changes already since `v3.1.1`.

btw, in terms of how we define what each number in `v3.1.2` really means, we could follow the *semantic versioning* standard or come up with whatever works best for us. 

See [wiki](https://github.com/mars-sim/mars-sim/wiki/Other-Ideas).

In summary, 

```
Given a version number MAJOR.MINOR.PATCH, increment the:

MAJOR version when you make incompatible API changes,
MINOR version when you add functionality in a backwards compatible manner, and
PATCH version when you make backwards compatible bug fixes.
Additional labels for pre-release and build metadata are available as extensions to the MAJOR.MINOR.PATCH format.
```

Q: Any comments ?

**Comment by bevans2000** [comment]
Yes that makes sense. I'm busy this weekend but will have a look when I get a chance. I will tag a y issues I think are needed for 3.1.2

**Reply by mokun** [reply]
So let's see if we can align what we are doing to this scheme `MAJOR.MINOR.PATCH`

Since we're transitioning from `3.1.1` to `3.1.2`, we should essentially be doing *patch*ing work only.

But we are in fact having been adding many new *functionality*. 

So, how about I change all the version description to `3.2.1`, so that it reflects on what we have been doing !?

**Reply by mokun** [reply]
> So, how about I change all the version description to 3.2.1, so that it reflects on what we have been doing !?

Or rather, we should change all version tags to to `3.2.0`

**Reply by mokun** [reply]
> > So, how about I change all the version description to 3.2.1, so that it reflects on what we have been doing !?
> 
> Or rather, we should change all version tags to to `3.2.0`

So, I'm going to delete `3.1.2` milestone, since we have been in fact doing `MINOR` changes and not `PATCH` changes.

**Reply by mokun** [reply]
Here, in this [milestones](https://github.com/mars-sim/mars-sim/milestones) page, we have 

![image](https://user-images.githubusercontent.com/1168584/122650171-3975a980-d0e6-11eb-82cd-48a955dcd993.png)

- 16 issues closed under `3.1.2`.
- 9 issues still open under `3.1.2`

When we delete `3.1.2`, how about we transfer ALL those closed ones to `3.2.0`.

But can we decide on which of those open ones we truly want to be in `3.2.0` or in future versions such as `3.2.1`, `3.2.2`, `3.3.0`, etc. ?

**Reply by bevans2000** [reply]
Yes transfer them to 3.2.0 I'm ok with that. Of course if you want to implement true release management then for the patch release we will have to use a path/bugfix branch for any patch release.

**Reply by mokun** [reply]
> implement true release management then for the patch release we will have to use a path/bugfix branch for any patch release.

Yes. Can you show me how to transition into doing that the proper way in git ?

**Reply by mokun** [reply]
Guess it entails committing to `3.2.0` if the work has to do with `3.2.0`, instead of the old way of committing everything to `Master` branch, right ?

**Reply by bevans2000** [reply]
It all depends whether you want to follow Git Flow or Trunk development. For a small team trunk is better as it js simpler. In a nutshell everyone commits to trunk (master branch in git speak). Automated CI runs after every commit, we have this now. At release time create a new Git Tag on master to snapshot the source at that point. Next release development continues in trunk.
If a patch is needed then create a release branch named after the release anchored to the release rag. The branch is an exact snapshot of the release to be fixed so you know the patch can be very focused on just the problem. Make the patch release, make a new tag and then merge the release branch back to trunk.
Then for big changes, e,g. a big refactoring,  you can use a feature branch. This allows destructive changes in isolation without breaking the trunk build. Breaking the trunk build is a cardinal sin in trunk development.
There is lots of info online but at the end of the day  you should adopt an approach that works for the team, whatever that is.

**Reply by mokun** [reply]
> Git Flow or Trunk development. For a small team trunk is better as it js simpler. In a nutshell everyone commits to trunk (master branch in git speak). 

Sure.  May be for now, it's good enough we stick with the trunk-based development since it's just a matter us two coordinating things together.

**Reply by mokun** [reply]
> Yes transfer them to 3.2.0 I'm ok with that. Of course if you want to implement true release management then for the patch release we will have to use a path/bugfix branch for any patch release.

I've just run a command-line based tool called `VersionTag` (located in `mars-sim-main`) to update all the class files with v3.2.0 and new date. 

Just uploaded the [commit](https://github.com/mars-sim/mars-sim/commit/900a62fac1d8bb5c7a5dc07829f40007763ceeb5).

**Reply by bevans2000** [reply]
You don't need to run VersionTag. When you create a Release in GitHub it creates a tag for you. You can see tag 5773 was created for release 3.1.2. I will clone the tag with a more meaningful name v3.1.2 later.

**Reply by mokun** [reply]
> When you create a Release in GitHub it creates a tag for you. You can see tag 5773 was created for release 3.1.2.

Sure. We've been using GitHub to create a bunch of *unofficial releases* once every two/three months or so. That I don't worry. It's just that we need to decide on whether we want to continue on the tradition of using the version header on each class as in the past or not.

**Comment by mokun** [comment]
So, let's do the v3.2.0 release in a week or so. Say our drop dead date in on or before 3 Jul 2021 ?

Seems there are always bugs to fix and, especially when we have to refactor and when we add new features. 

In future, I'm thinking we should roughly

1. Limit to add no more than 2 brand new major features in one major release. e.g. v3.2.0 to v4.0.0

2. Limit to add no more than 3 brand new minor features in one minor release. e.g. v3.2.2 to v3.2.0

3. Limit to add no more than 10 patches in one minor release. e.g. v.3.2.2 to v3.2.3

https://www.geeksforgeeks.org/introduction-semantic-versioning/

**Comment by mokun** [comment]
@bevans2000 ,

We still have 9 open issues in [3.2.0](https://github.com/mars-sim/mars-sim/milestones/3.2.0)

Can we move some of them to 3.2.1 ? 

let's set the goal of releasing the 3.2.0 on July 4. 

Say we freeze the change by no later than Jul 4, 2021 1400 PDT. 

Then after Jul 5, 2021 0000 PDT, we can resume our fixes and add new features on 3.2.1 or after. Shall we ?

**Reply by bevans2000** [reply]
Yes that seems reasonable. I don't have much spare time at the moment so won't be making any other big changes.

**Reply by mokun** [reply]
Is there anything else you want to wrap up before we freeze it ? 

I'm still looking at the use of Drone for Delivery Mission. But it seems that even the normal Trading mission cannot be started.

**Comment by bevans2000** [comment]
I will have a look at the Delivery Mission tomorrow.

**Reply by mokun** [reply]
@bevans2000 ,

I'll do one more commit in a moment.

Do you have one more commit to make before freezing it ?

**Reply by mokun** [reply]
> I will have a look at the Delivery Mission tomorrow.

Okay.

**Reply by bevans2000** [reply]
No I did the things I wanted. Next I am going to look at #355 and #359.

**Reply by mokun** [reply]
> No I did the things I wanted. Next I am going to look at #355 and #359.

I still need to drop one more or so to fix up the time ratio, which is a critical issue.

**Reply by mokun** [reply]
> I still need to drop one more or so to fix up the time ratio, which is a critical issue.

@bevans2000 , 

Can you check if the sim runs smoothly on your laptop after this [commit](https://github.com/mars-sim/mars-sim/commit/b02b789ab26f777e702b27ee4c5d67e134a3385c) ?

**Comment by mokun** [comment]
@bevans2000 ,

I compiled two binaries : 
`3.2.0_remote_java11.jar`
`3.2.0_swing_java11.jar`

Both of them have the following issues with `-Xms256m`.


```
Microsoft Windows [Version 10.0.19043.1052]
(c) Microsoft Corporation. All rights reserved.

D:\Data\msp binary\v3.2.0>java -jar 3.2.0_remote_java11.jar
      JAVA_HOME : C:\Program Files\BellSoft\LibericaJDK-11\
 File.separator : \
      JAVA_HOME : "C:\Program Files\BellSoft\LibericaJDK-11\
   Java Command : "C:\Program Files\BellSoft\LibericaJDK-11\bin\java"
Command: "C:\Program Files\BellSoft\LibericaJDK-11\bin\java" --illegal-access=deny --add-opens java.desktop/com.sun.java.swing.plaf.windows=ALL-UNNAMED -Djava.util.logging.config.file=logging.properties -cp .;*;jars\* org.mars_sim.headless.MarsProjectHeadless -Xms256m -Xmx1536m -new
2021-07-03T19:00:51.1 (Config) MarsProjectHeadless : Starting Mars Simulation Project v3.2.0 - Build 5900 - 64-bit Java 11.0.5-BellSoft - 8 CPU threads
2021-07-03T19:00:51.1 (Config) MarsProjectHeadless : List of input args : [Ljava.lang.String;@30a3107a
Problem with arguments: Unrecognized option: -Xms256m
usage: [for mars-sim headless edition]
 -datadir <path to data directory>   Path to the data directory for
                                     simulation files (defaults to
                                     user.home)
 -help                               Help of the options
 -lat <latitude>                     Set the latitude of the new template
                                     Settlement
 -load <path to simulation file>     Load the a previously saved sim,
                                     default is used if none specifed
 -lon <longitude>                    Set the longitude of the new template
                                     Settlement
 -new                                Create a new simulation if one is not
                                     present
 -noremote                           Do not start a remote console service
 -remote <port number>               Run the remote console service
                                     [default]
 -resetadmin                         Reset the internal admin password
 -sponsor <sponsor>                  Set the sponsor for the settlement
                                     template
 -template <template>                New simulation from a template
 -timeratio <Ratio (power of 2)>     Define the time ratio of the
                                     simulation

D:\Data\msp binary\v3.2.0>java -jar 3.2.0_remote_java11.jar -new
      JAVA_HOME : C:\Program Files\BellSoft\LibericaJDK-11\
 File.separator : \
      JAVA_HOME : "C:\Program Files\BellSoft\LibericaJDK-11\
   Java Command : "C:\Program Files\BellSoft\LibericaJDK-11\bin\java"
Command: "C:\Program Files\BellSoft\LibericaJDK-11\bin\java" --illegal-access=deny --add-opens java.desktop/com.sun.java.swing.plaf.windows=ALL-UNNAMED -Djava.util.logging.config.file=logging.properties -cp .;*;jars\* org.mars_sim.headless.MarsProjectHeadless -Xms256m -Xmx1536m -new
2021-07-03T19:01:26.7 (Config) MarsProjectHeadless : Starting Mars Simulation Project v3.2.0 - Build 5900 - 64-bit Java 11.0.5-BellSoft - 8 CPU threads
2021-07-03T19:01:26.7 (Config) MarsProjectHeadless : List of input args : [Ljava.lang.String;@6ee52dcd
Problem with arguments: Unrecognized option: -Xms256m
usage: [for mars-sim headless edition]
 -datadir <path to data directory>   Path to the data directory for
                                     simulation files (defaults to
                                     user.home)
 -help                               Help of the options
 -lat <latitude>                     Set the latitude of the new template
                                     Settlement
 -load <path to simulation file>     Load the a previously saved sim,
                                     default is used if none specifed
 -lon <longitude>                    Set the longitude of the new template
                                     Settlement
 -new                                Create a new simulation if one is not
                                     present
 -noremote                           Do not start a remote console service
 -remote <port number>               Run the remote console service
                                     [default]
 -resetadmin                         Reset the internal admin password
 -sponsor <sponsor>                  Set the sponsor for the settlement
                                     template
 -template <template>                New simulation from a template
 -timeratio <Ratio (power of 2)>     Define the time ratio of the
                                     simulation

D:\Data\msp binary\v3.2.0>java -jar 3.2.0_swing_java11.jar -new
      JAVA_HOME : C:\Program Files\BellSoft\LibericaJDK-11\
 File.separator : \
      JAVA_HOME : "C:\Program Files\BellSoft\LibericaJDK-11\
   Java Command : "C:\Program Files\BellSoft\LibericaJDK-11\bin\java"
Command: "C:\Program Files\BellSoft\LibericaJDK-11\bin\java" --add-opens java.base/java.util=ALL-UNNAMED --add-opens java.base/java.text=ALL-UNNAMED --add-opens java.base/java.lang.reflect=ALL-UNNAMED --add-opens java.base/java.net=ALL-UNNAMED --add-opens java.base/java.lang=ALL-UNNAMED --add-opens java.base/jdk.internal.loader=ALL-UNNAMED --add-opens java.desktop/javax.swing=ALL-UNNAMED --add-opens java.desktop/javax.swing.text=ALL-UNNAMED --add-opens java.desktop/java.awt.font=ALL-UNNAMED --add-opens java.desktop/java.awt.geom=ALL-UNNAMED --add-opens java.desktop/java.awt=ALL-UNNAMED --add-opens java.desktop/java.beans=ALL-UNNAMED --add-opens java.desktop/javax.swing.table=ALL-UNNAMED --add-opens java.desktop/com.sun.awt=ALL-UNNAMED --add-opens java.desktop/sun.awt=ALL-UNNAMED --add-opens java.desktop/sun.swing=ALL-UNNAMED --add-opens java.desktop/sun.font=ALL-UNNAMED --add-opens java.desktop/javax.swing.plaf.basic=ALL-UNNAMED --add-opens java.desktop/javax.swing.plaf.synth=ALL-UNNAMED --add-opens java.desktop/com.sun.java.swing.plaf.windows=ALL-UNNAMED -Djava.util.logging.config.file=logging.properties -cp .;*;jars\* org.mars_sim.main.MarsProject -Xms256m -Xmx1536m -new
2021-07-03T19:01:43.6 (Config) MarsProject : Starting Mars Simulation Project v3.2.0 - Build 5900 - 64-bit Java 11.0.5-BellSoft - 8 CPU threads
Problem with arguments: Unrecognized option: -Xms256m
usage: [for mars-sim edition]
 -datadir <path to data directory>   Path to the data directory for
                                     simulation files (defaults to
                                     user.home)
 -help                               Help of the options
 -lat <latitude>                     Set the latitude of the new template
                                     Settlement
 -load <path to simulation file>     Load the a previously saved sim,
                                     default is used if none specifed
 -lon <longitude>                    Set the longitude of the new template
                                     Settlement
 -new                                Create a new simulation if one is not
                                     present
 -noaudio                            Disable the audio
 -nogui                              Disable the main UI
 -sponsor <sponsor>                  Set the sponsor for the settlement
                                     template
 -template <template>                New simulation from a template
 -timeratio <Ratio (power of 2)>     Define the time ratio of the
                                     simulation

D:\Data\msp binary\v3.2.0>
```

**Reply by mokun** [reply]
Seems like the following two classes need to be updated to match up with the new changes made in `MarsProject` and `MarsProjectHeadless` : 
- MarsProjectStarter
- MarsProjectHeadlessStarter

**Reply by mokun** [reply]
I'm making changes to `MarsProjectStarter` and `MarsProjectHeadlessStarter` as we speak.

**Reply by bevans2000** [reply]
What do these 2 Starter classes actually do ?

**Reply by mokun** [reply]
They are preprocessing the command stream when running the compiled jar. They pass the arguments onto MarsProject or MarsProjectHeadless.

**Reply by bevans2000** [reply]
So your problem with '-Xmx' is because in the Starter class you are putting this as an argument "after" the Jar/class file. The '-Xmx' is actually an 'option' and should appear "before" the class/jar file. (https://docs.oracle.com/javase/7/docs/technotes/tools/windows/java.html) These option do not get passed to the application but are handled by the JVM. You just need to change how the command line is constructed. To be honest I can't see what they provide extra to a basic 'sh' or 'bat' file wrapper like many Java based application. To get the Starter running the user must have got a Java process running which is normally 80% of the battle.

**Reply by mokun** [reply]
@bevans2000 

> So your problem with '-Xmx' is because in the Starter class you are putting this as an argument "after" the Jar/class file. The '-Xmx' is actually an 'option' and should appear "before" the class/jar file. (https://docs.oracle.com/javase/7/docs/technotes/tools/windows/java.html) These option do not get passed to the application but are handled by the JVM. You just need to change how the command line is constructed

Thanks for pointing that `-xmx1536M` is actually one of those options and not arguments.

On top of that, in both starter classes, the classpath option needs to be fixed.

```
        // The following few appends create this option string that is being used up to v3.07
        // -cp .;*;jars\*
        command.append(" -cp .")
    	.append(File.pathSeparator) // File.pathSeparator is a semi-colon ';' in windows or a colon ':' in linux
    	.append("*")
    	.append(File.pathSeparator) // File.pathSeparator is a semi-colon ';' in windows or a colon ':' in linux
    	.append("jars")
    	.append(File.separator)     // File.separator is a backslash '\' in windows or a forward slash '/' in linux
    	.append("*");
    // v3.1.0 started distributing a single jar binary, instead of having a hierarchy of folders and files.

```

Currently,  when I want to start a jar (say, 5913.jar) in the midst of many jar files in the same folder, I won't be able to do so.

![image](https://user-images.githubusercontent.com/1168584/124642149-5c24f380-de44-11eb-953b-bc48865d60be.png)

It will load up 5913.jar initially but will search for all the jar having the main class. So it will first go to 5910.jar and load up its MarsProject's main() up, instead of 5913.jar's main().

Q: is there a way to fix this to allow multiple jar files in one folder while picking only the specified jar file ? 

I know this command sequence is to blame :
 
>  -cp .;*;jars\*

In v3.07 distribution, things are structured with both src and binary files and with folders nicely. So it will go into the jars folder to look for all other jars.

I recalled commenting that to @shirishag75 in the past that he's not supposed to place a bunch of jar files of different builds together in one folder or else it will default to running the first sorted alphabetical jar.

**Reply by mokun** [reply]
Guess the jar file comes AFTER the option and there's no way for Java to know the jar filename until after it processes the filename.

Having said that, i think we should distribute the binary already well placed inside a jar folder. Then it will sort of minimize the user from accidentally placing more than one start-able jar file inside.

**Reply by bevans2000** [reply]
Maybe I'm missing something but why not just build a single executable JAR file named with the release number? Then there is no need for a class path argument nor the Starter class.
Also a true release package should have the release embedded in the folder name so by default it only coNtains one release. Then include a starter shell script or bat file that would start the simulation. This pattern is used  by lots of Java applications.

**Reply by mokun** [reply]
> Maybe I'm missing something but why not just build a single executable JAR file named with the release number? Then there is no need for a class path argument nor the Starter class.

Sure. After v3.07 we start going toward this approach, instead of relying on the the starter class.

To complete the transition, we also need to configure the logging files since it was done in the starter class.

> Also a true release package should have the release embedded in the folder name so by default it only coNtains one release. Then include a starter shell script or bat file that would start the simulation. This pattern is used by lots of Java applications.

Yes, I agree we should use the batch file, which is way more flexible. So if we make any mistakes after releasing the official release version, we can simply upload a fixed batch file for users to use.

**Reply by mokun** [reply]
So after releasing v3.2.0, we'll discontinue the use of the starter classes and also work on optimizing the saving of those logging files . Agree ?

Since not every user prefer to start mars-sim in a command prompt, it's a good thing to ask users to look for the log files saved in their machine and post these log files if the simulation ever crashed. Without that, it's harder to debug, you know.

**Reply by bevans2000** [reply]
Sounds good.

**Comment by mokun** [comment]
I removed `-Xms256m` in build 5901.

But turns out it also doesn't like `Xmx1536m` for some strange reasons.

When I try 
> java -X 

The two options are still there.

```
D:\Data\msp binary\v3.2.0>java -X

    -Xbatch           disable background compilation
    -Xbootclasspath/a:<directories and zip/jar files separated by ;>
                      append to end of bootstrap class path
    -Xcheck:jni       perform additional checks for JNI functions
    -Xcomp            forces compilation of methods on first invocation
    -Xdebug           provided for backward compatibility
    -Xdiag            show additional diagnostic messages
    -Xfuture          enable strictest checks, anticipating future default
    -Xint             interpreted mode execution only
    -Xinternalversion
                      displays more detailed JVM version information than the
                      -version option
    -Xloggc:<file>    log GC status to a file with time stamps
    -Xmixed           mixed mode execution (default)
    -Xmn<size>        sets the initial and maximum size (in bytes) of the heap
                      for the young generation (nursery)
    -Xms<size>        set initial Java heap size
    -Xmx<size>        set maximum Java heap size
    -Xnoclassgc       disable class garbage collection
    -Xrs              reduce use of OS signals by Java/VM (see documentation)
    -Xshare:auto      use shared class data if possible (default)
    -Xshare:off       do not attempt to use shared class data
    -Xshare:on        require using shared class data, otherwise fail.
    -XshowSettings    show all settings and continue
    -XshowSettings:all
                      show all settings and continue
    -XshowSettings:locale
                      show all locale related settings and continue
    -XshowSettings:properties
                      show all property settings and continue
    -XshowSettings:vm
                      show all vm related settings and continue
    -XshowSettings:system
                      (Linux Only) show host system or container
                      configuration and continue
    -Xss<size>        set java thread stack size
    -Xverify          sets the mode of the bytecode verifier
    --add-reads <module>=<target-module>(,<target-module>)*
                      updates <module> to read <target-module>, regardless
                      of module declaration.
                      <target-module> can be ALL-UNNAMED to read all unnamed
                      modules.
    --add-exports <module>/<package>=<target-module>(,<target-module>)*
                      updates <module> to export <package> to <target-module>,
                      regardless of module declaration.
                      <target-module> can be ALL-UNNAMED to export to all
                      unnamed modules.
    --add-opens <module>/<package>=<target-module>(,<target-module>)*
                      updates <module> to open <package> to
                      <target-module>, regardless of module declaration.
    --illegal-access=<value>
                      permit or deny access to members of types in named modules
                      by code in unnamed modules.
                      <value> is one of "deny", "permit", "warn", or "debug"
                      This option will be removed in a future release.
    --limit-modules <module name>[,<module name>...]
                      limit the universe of observable modules
    --patch-module <module>=<file>(;<file>)*
                      override or augment a module with classes and resources
                      in JAR files or directories.
    --disable-@files  disable further argument file expansion
    --source <version>
                      set the version of the source in source-file mode.

These extra options are subject to change without notice.
```

**Reply by mokun** [reply]
Also, I wanna run `MarsProject -html.launch` eclipse launcher to generate a new set of html files since I recalled we did change a few manufacturing processes as well as having the delivery drone added as a new vehicle.

However, I get the following error : 

```
2021-07-03T22:41:28.0 (Config) MarsProject : Starting Mars Simulation Project v3.2.0 - Build 5901 - 64-bit Java 13.0.2 - 8 CPU threads
Problem with arguments: Unrecognized option: -html
usage: [for mars-sim edition]
 -datadir <path to data directory>   Path to the data directory for
                                     simulation files (defaults to
                                     user.home)
 -help                               Help of the options
 -lat <latitude>                     Set the latitude of the new template
                                     Settlement
 -load <path to simulation file>     Load the a previously saved sim,
                                     default is used if none specifed
 -lon <longitude>                    Set the longitude of the new template
                                     Settlement
 -new                                Create a new simulation if one is not
                                     present
 -noaudio                            Disable the audio
 -nogui                              Disable the main UI
 -sponsor <sponsor>                  Set the sponsor for the settlement
                                     template
 -template <template>                New simulation from a template
 -timeratio <Ratio (power of 2)>     Define the time ratio of the
                                     simulation

```

**Reply by mokun** [reply]
@bevans2000 

I also discovered that if I load the Config Editor (aka Site Editor) before starting a sim, then as it loads up the Main Window, it would crash and the simulation quits. 

![image](https://user-images.githubusercontent.com/1168584/124377550-42778680-dc61-11eb-8dc3-c948454a7cad.png)

**Reply by bevans2000** [reply]
Strange as I checked this. I can look at it later.

**Reply by bevans2000** [reply]
The -html option is fixed

**Reply by bevans2000** [reply]
I have fixed the -html problem

**Reply by mokun** [reply]
> I have fixed the -html problem

Thanks !

**Reply by mokun** [reply]
> @bevans2000
> 
> I also discovered that if I load the Config Editor (aka Site Editor) before starting a sim, then as it loads up the Main Window, it would crash and the simulation quits.
> 

I wonder what could have gone wrong in MarsProject that triggers the Exception and thereby calling `exitWithError` to create that popup JOptionPane box ?

```
		// Do it
		try {
			if (useGUI) {
				MainWindow.startSplash();
				// System.setProperty("sun.java2d.opengl", "true"); // not compatible with
				// SplashWindow and SimulationConfigEditor
				if (!MainWindow.OS.contains("linux")) {
					System.setProperty("sun.java2d.ddforcevram", "true"); // question: is this compatible with opengl in
																		// linux and macos ?
				}
			}

			// Preload the Config
			SimulationConfig.instance().loadConfig();
			
			// Get user choices if there is no template defined or a preload
			if ((builder.getTemplate() == null) && (builder.getSimFile() == null)) {
				logger.config("Please go to the mars-sim console's Main Menu to choose an option.");
				
				int type = interactiveTerm.startConsoleMainMenu();
				if (type == 1) {
					// Start sim config editor
					SimulationConfigEditor editor = new SimulationConfigEditor(SimulationConfig.instance());
					logger.config("Running the site editor...");
					editor.waitForCompletion();
				}
			
				else if (type == 2) {
					// Load simulation
					String filePath = selectSimFile(false);
					if (filePath != null) {
						builder.setSimFile(filePath);
					}
				}
			}
			
			// Build and run the simulator
			builder.start();
			
			// Start the wait layer
			InteractiveTerm.startLayer();
			
			// Start beryx console
			startConsoleThread();
			
			if (useGUI) {
				setupMainWindow(false);
			} 
		}
		catch(Exception e) {
			// Catch everything
			exitWithError("Problem starting " + e.getMessage(), e);

		}


	/**
	 * Exit the simulation with an error message.
	 * 
	 * @param message the error message.
	 * @param e       the thrown exception or null if none.
	 */
	private void exitWithError(String message, Exception e) {
		if (e != null) {
			logger.log(Level.SEVERE, message, e);
		} else {
			logger.log(Level.SEVERE, message);
		}

		if (useGUI) {
			JOptionPane.showMessageDialog(null, message, "Error", JOptionPane.ERROR_MESSAGE);
		}
		System.exit(1);
	}
```

**Reply by bevans2000** [reply]
I have found the problem. I cut down on the excessive use of Threads/Executors to start the UI to improve the start-up. But the window constructors are littered with use of SwingUtilities.invokeLater. This means that sometimes the internals of the windows have not been created. There is no need to use invokeLater when building the windows because it can introduce race conditions.
 I will look into fixing it, might not be tonight as I've just had a long training run so mentally a bit tired.

**Reply by bevans2000** [reply]
Strike that. I think it's fixed in this commit 8810b287e0e6d64cb6fd1f5020f851db383e35e3

**Reply by mokun** [reply]
> Strike that. I think it's fixed in this commit [8810b28](https://github.com/mars-sim/mars-sim/commit/8810b287e0e6d64cb6fd1f5020f851db383e35e3)

That's great !

**Reply by mokun** [reply]
> the window constructors are littered with use of SwingUtilities.invokeLater. This means that sometimes the internals of the windows have not been created.

But i heard it's better of building swing components inside swing ui thread, is that true ?

Also, constructing weblaf's components sometimes require to be run inside swing ui or else it won't work.

**Reply by bevans2000** [reply]
I can't comment on weblaf. But if the objects  are created via the thread then care needs to  be taken before they can be referenced because using the swing ui thread you don't know when they exactly get created.

**Reply by mokun** [reply]
> `-Xms256m` in build 5901.
> 
> But turns out it also doesn't like `Xmx1536m`

How do I tell MarsProject or package org.apache.commons.cli to ignore these two java specific arguments, namely, `-Xms256m` and Xmx1536m`?

They should have been "absorbed" by the java command but I can't make it go away in MarsProjectStarter.

When starting a jar, I believe it first reads the main() method in MarsProjectStarter. 

This is the default setting compiled under the maven.
